src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTarget.java

```diff
@@ -2986,7 +2986,50 @@ public void withSavedDoc(Command onsaved)
    {
       docUpdateSentinel_.withSavedDoc(onsaved);
    }
-   
+
+
+   // sadly due to Java's memory model we can't modify local variables
+   // inside of a lambda but class variables are protected and
+   // are one of the recommended workarounds
+   private int totalWords = 0;
+   private int selectionWords = 0;
+
+   @Handler
+   void onWordCount()
+   {
+      totalWords = 0;
+      selectionWords = 0;
+
+      Position pos = docDisplay_.getCursorPosition();
+      TextFileType fileType = docDisplay_.getFileType();
+      Iterable<Range> wordSource = docDisplay_.getWords(
+         fileType.getTokenPredicate(),
+         docDisplay_.getFileType().getCharPredicate(),
+         pos,
+         null);
+
+      totalWords = 0;
+      wordSource.forEach((r) -> totalWords++);
+
+      Range selectionRange = docDisplay_.getSelectionRange();
+      if (selectionRange.getStart().compareTo(selectionRange.getEnd()) != 0)
+      {
+         wordSource = docDisplay_.getWords(
+            fileType.getTokenPredicate(),
+            docDisplay_.getFileType().getCharPredicate(),
+            docDisplay_.getSelectionStart(),
+            docDisplay_.getSelectionEnd());
+
+         selectionWords = 0;
```

Comment:
How does this variable get reset between command invocations if no selection is present?

Reply:
It was being reset above, this was leftover code from before I moved it.